"""Test cases for UXC message encoding/decoding."""

import queue
import unittest
from typing import Optional, Union

from bitkey import uxc
from bitkey_proto import wallet_pb2


class TestUXC(unittest.TestCase):
    """Test cases for UXC encoding/decoding."""

    def setUp(self):
        """Pre-task setup."""

        def _read(q: queue.Queue) -> callable:

            def __read(num_bytes: int) -> Optional[Union[bytes, bytearray]]:
                try:
                    return q.get(block=True, timeout=0.001)
                except queue.Empty:
                    return None

            return __read

        self._recv_q: queue.Queue = queue.Queue()
        self._send_q: queue.Queue = queue.Queue()

        self.host: uxc.UXCHost = uxc.UXCHost(write=self._send_q.put, read=_read(self._recv_q))
        self.device: uxc.UXCDevice = uxc.UXCDevice(write=self._recv_q.put, read=_read(self._send_q))

        self.host.start()
        self.device.start()

    def tearDown(self):
        """Post-task teardown."""

        self.host.stop()
        self.device.stop()

    def test_compute_crc(self):
        """Tests computing a CRC for a message."""
        payload = [
            0x49, 0x27, 0x6D, 0x20, 0x77, 0x6F, 0x72, 0x72, 0x69, 0x65, 0x64,
            0x20, 0x74, 0x68, 0x61, 0x74, 0x20, 0x74, 0x68, 0x65, 0x20, 0x62,
            0x61, 0x62, 0x79, 0x20, 0x74, 0x68, 0x69, 0x6E, 0x6B, 0x73, 0x20,
            0x70, 0x65, 0x6F, 0x70, 0x6C, 0x65, 0x20, 0x63, 0x61, 0x6E, 0x27,
            0x74, 0x20, 0x63, 0x68, 0x61, 0x6E, 0x67, 0x65, 0x2E,
        ]

        hdr = uxc.UXCMsgHdr()
        hdr.proto_tag = 0x1337
        hdr.send_seq_num = 0x13
        hdr.ack_seq_num = 0x7
        hdr.flags = 0x1
        hdr.payload_len = len(payload)

        self.assertEqual(48783, uxc.UXCMsg.compute_crc(hdr, payload))

    def test_msg_from_bytes_invalid(self):
        """Tests the cases where decoding a message should fail."""
        with self.assertRaises(uxc.UXCError):
            _ = uxc.UXCMsg.from_bytes(b"\x01\x02")

        with self.assertRaises(uxc.UXCError):
            hdr = uxc.UXCMsgHdr()
            hdr.payload_len = 123
            _ = uxc.UXCMsg.from_bytes(bytes(hdr))

        with self.assertRaises(uxc.UXCError):
            hdr = uxc.UXCMsgHdr()
            hdr.payload_len = 1
            hdr.crc = 1337
            _ = uxc.UXCMsg.from_bytes(bytes(hdr) + b"\x02")

    def test_host_device(self):
        """Tests sending multiple messages from host to device."""

        self.assertTrue(self.device.send_empty())
        _ = self.host.get_empty()

        cert = bytearray([
            0x57, 0x65, 0x6C, 0x6C, 0x2C, 0x20, 0x69, 0x74, 0x20, 0x64, 0x6F,
            0x65, 0x73, 0x2C, 0x20, 0x62, 0x65, 0x63, 0x61, 0x75, 0x73, 0x65,
            0x20, 0x74, 0x68, 0x61, 0x74, 0x20, 0x61, 0x6D, 0x6F, 0x75, 0x6E,
            0x74, 0x20, 0x69, 0x73, 0x20, 0x63, 0x61, 0x6C, 0x6C, 0x65, 0x64,
            0x20, 0x6D, 0x79, 0x20, 0x71, 0x75, 0x6F, 0x74, 0x65, 0x2E, 0x20,
            0x54, 0x68, 0x61, 0x74, 0x27, 0x73, 0x20, 0x6D, 0x79, 0x20, 0x72,
            0x61, 0x74, 0x65, 0x2E, 0x20, 0x53, 0x6F, 0x20, 0x74, 0x68, 0x65,
            0x20, 0x6E, 0x65, 0x78, 0x74, 0x20, 0x66, 0x69, 0x6C, 0x6D, 0x20,
            0x49, 0x27, 0x6D, 0x20, 0x6F, 0x66, 0x66, 0x65, 0x72, 0x65, 0x64,
            0x2C, 0x20, 0x74, 0x68, 0x65, 0x79, 0x20, 0x68, 0x61, 0x76, 0x65,
            0x20, 0x74, 0x6F, 0x20, 0x70, 0x61, 0x79, 0x20, 0x74, 0x68, 0x61,
            0x74, 0x20, 0x73, 0x61, 0x6D, 0x65, 0x20, 0x61, 0x6D, 0x6F, 0x75,
            0x6E, 0x74, 0x2E, 0x20, 0x45, 0x76, 0x65, 0x6E, 0x20, 0x69, 0x66,
            0x20, 0x49, 0x20, 0x64, 0x6F, 0x20, 0x61, 0x20, 0x62, 0x61, 0x64,
            0x20, 0x6A, 0x6F, 0x62, 0x2E, 0x20, 0x54, 0x68, 0x61, 0x74, 0x20,
            0x6D, 0x65, 0x61, 0x6E, 0x73, 0x2C, 0x20, 0x61, 0x73, 0x20, 0x6C,
            0x6F, 0x6E, 0x67, 0x20, 0x61, 0x73, 0x20, 0x49, 0x27, 0x6D, 0x20,
            0x6F, 0x66, 0x66, 0x65, 0x72, 0x65, 0x64, 0x20, 0x65, 0x76, 0x65,
            0x6E, 0x20, 0x6F, 0x6E, 0x65, 0x20, 0x6D, 0x6F, 0x72, 0x65, 0x20,
            0x6D, 0x6F, 0x76, 0x69, 0x65, 0x2C, 0x20, 0x49, 0x20, 0x63, 0x6F,
            0x75, 0x6C, 0x64, 0x20, 0x67, 0x65, 0x74, 0x20, 0x74, 0x77, 0x6F,
            0x20, 0x6D, 0x6F, 0x72, 0x65, 0x20, 0x6D, 0x69, 0x6C, 0x2E, 0x20,
            0x45, 0x76, 0x65, 0x6E, 0x20, 0x69, 0x66, 0x20, 0x49, 0x20, 0x64,
            0x6F, 0x20, 0x61, 0x20, 0x62, 0x61, 0x64, 0x20, 0x6A, 0x6F, 0x62,
            0x2C, 0x20, 0x74, 0x68, 0x65, 0x79, 0x27, 0x76, 0x65, 0x20, 0x67,
            0x6F, 0x74, 0x20, 0x74, 0x6F, 0x20, 0x67, 0x69, 0x76, 0x65, 0x20,
            0x6D, 0x65, 0x20, 0x74, 0x68, 0x61, 0x74, 0x20, 0x6F, 0x74, 0x68,
            0x65, 0x72, 0x20, 0x74, 0x77, 0x6F, 0x20, 0x6D, 0x69, 0x6C, 0x2E
        ])
        self.assertTrue(self.device.send_cert(cert))

        received_cert = self.host.get_cert(cert_type=wallet_pb2.cert_get_cmd.cert_type.BATCH_CERT)
        self.assertEqual(cert, received_cert)

    def test_empty_cmd_rsp(self):
        """Test for empty message exchange between host and device."""

        self.assertTrue(self.device.send_empty())
        self.assertTrue(self.host.get_empty())

    def test_mfgtest_gpio_cmd_rsp(self):
        """Tests for the GPIO manufacturing messages."""

        self.assertTrue(self.device.send_gpio_status(0))
        self.assertEqual(0, self.host.gpio_read(1, 7))

        self.assertTrue(self.device.send_gpio_status(1))
        self.assertEqual(1, self.host.gpio_set(1, 7))

        self.assertTrue(self.device.send_gpio_status(0))
        self.assertEqual(0, self.host.gpio_clear(1, 7))

    def test_send_recv_events(self):
        """Tests sending and receiving bitlogs."""

        expected_events = [
            0x49, 0x20, 0x66, 0x65, 0x65, 0x6C, 0x20, 0x6C, 0x69, 0x6B, 0x65,
            0x20, 0x79, 0x6F, 0x75, 0x27, 0x72, 0x65, 0x20, 0x6A, 0x75, 0x73,
            0x74, 0x20, 0x68, 0x65, 0x72, 0x65, 0x20, 0x66, 0x6F, 0x72, 0x20,
            0x74, 0x68, 0x65, 0x20, 0x7A, 0x69, 0x70, 0x6C, 0x69, 0x6E, 0x65,
            0x2E
        ]
        expected_version = 1

        events = expected_events[:]
        remaining_bytes = len(expected_events)
        while remaining_bytes > 0:
            chunk_size = min(10, remaining_bytes)
            chunk = events[: chunk_size]

            events = events[chunk_size :]
            remaining_bytes -= chunk_size

            self.assertTrue(self.device.send_event_data(bytes(chunk), remaining_bytes, expected_version))

        actual_version, actual_events = self.host.get_events()
        self.assertEqual(expected_version, actual_version)
        self.assertEqual(bytes(expected_events), actual_events)


if __name__ == "__main__":
    unittest.main()
